src/Documents/Commands.ts

```diff
@@ -0,0 +1,486 @@
+    constructor(commandList: IPrinterCommand[]) {
+        this._commands = commandList;
+    }
+
+    /** Display the commands that will be performed in a human-readable format. */
+    public showCommands(): string {
+        let result = '';
+        this._commands.forEach((c) => (result += `${c.toDisplay()}\n`));
+        return result;
+    }
+}
+
+/** A document of raw commands, ready to be sent to a printer. */
+export class CompiledDocument {
+    private rawCmdBuffer: Array<Uint8Array> = [];
+    private _commandLanugage: PrinterCommandLanguage;
+    get commandLanguage() {
+        return this._commandLanugage;
+    }
+
+    horizontalOffset = 0;
+    verticalOffset = 0;
+    lineSpacingDots = 5;
+
+    commandEffectFlags = PrinterCommandEffectFlags.none;
+
+    constructor(commandLang: PrinterCommandLanguage) {
+        this._commandLanugage = commandLang;
+    }
+
+    /**
+     * Gets a single buffer of the internal command set.
+     */
+    get commandBufferRaw(): Uint8Array {
+        const bufferLen = this.rawCmdBuffer.reduce((sum, arr) => sum + arr.byteLength, 0);
+        const buffer = new Uint8Array(bufferLen);
+        this.rawCmdBuffer.reduce((offset, arr) => {
+            buffer.set(arr, offset);
+            return arr.byteLength + offset;
+        }, 0);
+
+        return buffer;
+    }
+
+    /**
+     * Gets the text view of the command buffer. Do not send this to the printer, the encoding
+     * will break and commands will fail.
+     */
+    get commandBufferString(): string {
+        return new TextDecoder('ascii').decode(this.commandBufferRaw);
+    }
+
+    /** Add a raw command to the internal buffer. */
+    addRawCmd(array: Uint8Array): CompiledDocument {
+        this.rawCmdBuffer.push(array);
+        return this;
+    }
+
+    /** Clear the internal buffer. */
+    clearCommandBuffer(): CompiledDocument {
+        this.rawCmdBuffer = [];
+        return this;
+    }
+}
+
+/** A basic document builder, containing internal state to construct a document. */
+export abstract class DocumentBuilder implements IDocumentBuilder {
+    private _commands: IPrinterCommand[] = [];
+    protected _config: PrinterOptions;
+
+    constructor(config: PrinterOptions) {
+        this._config = config;
+    }
+
+    /** Gets the read-only config information */
+    get currentConfig() {
+        return structuredClone(this._config);
+    }
+
+    clear(): IDocumentBuilder {
+        this._commands = [];
+        return this;
+    }
+
+    showCommands(): string {
+        let result = '';
+        this._commands.forEach((c) => (result += `${c.name} - ${c.toDisplay()}\n`));
+        return result;
+    }
+
+    finalize(): IDocument {
+        return new Document(this._commands);
+    }
+
+    then(command: IPrinterCommand): IDocumentBuilder {
+        this._commands.push(command);
+        return this;
+    }
+}
+
+/** Flags to indicate special operations a command might cause. */
+export enum PrinterCommandEffectFlags {
```

Comment:
This is a footnote for a future feature that can provide some introspection and analysis on what a document is about to do. This could be used, for example, to show a warning indicating that a document is missing a print command and won't actually print anything. It snuck its way into this PR, it should have been kept to a separate branch.